atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/AsyncSessionManager.java

```diff
@@ -0,0 +1,321 @@
+                    (RemovalListener<UniqueCassandraCluster, CassandraClusterSessionPair>) (key, value, cause) -> {
+                        value.session().close();
+                        value.cluster().close();
+                    })
+            .build();
+
+    private final AtomicLong cassandraId = new AtomicLong();
+    // global for all sessions
+    private final AtomicLong sessionId = new AtomicLong();
+
+    private AsyncSessionManager(TaggedMetricRegistry taggedMetricRegistry) {
+        this.taggedMetricRegistry = taggedMetricRegistry;
+    }
+
+    private AsyncSessionManager() {
+        this.taggedMetricRegistry = null;
+    }
+
+    public AsyncClusterSession getAsyncSession(CassandraKeyValueServiceConfig config, Set<InetSocketAddress> servers) {
+        CassandraClusterSessionPair cassandraClusterSessionPair = getCassandraClusterSessionPair(config, servers);
+
+        long curId = sessionId.getAndIncrement();
+        String sessionName = cassandraClusterSessionPair.cluster().getClusterName()
+                + "-session" + (curId == 0 ? "" : "-" + curId);
+
+        ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                .setNameFormat(sessionName + "-%d")
+                .build();
+
+        return AsyncClusterSessionImpl.create(sessionName, cassandraClusterSessionPair, threadFactory);
+    }
+
+    public void closeClusterSession(AsyncClusterSession asyncClusterSession) {
+        // in this case nothing happens as the
+        if (!(asyncClusterSession instanceof AsyncClusterSessionImpl)) {
+            log.warn("Closing session which was not opened by this manager");
+        }
+    }
+
+    private CassandraClusterSessionPair getCassandraClusterSessionPair(CassandraKeyValueServiceConfig config,
+            Set<InetSocketAddress> servers) {
+        return clusters.get(ImmutableUniqueCassandraCluster.of(servers),
+                key -> createCassandraClusterSessionPair(config, servers)
+        );
+    }
+
+    private CassandraClusterSessionPair createCassandraClusterSessionPair(CassandraKeyValueServiceConfig config,
+            Set<InetSocketAddress> servers) {
+        Cluster cluster = createCluster(config, servers);
+        Session session = null;
+
+        try {
+            session = cluster.connectAsync().get();
+        } catch (Exception e) {
+            log.warn("Error on cluster connection");
+            Throwables.unwrapAndThrowAtlasDbDependencyException(e);
+        }
+
+        return ImmutableCassandraClusterSessionPair.of(cluster, Objects.requireNonNull(session));
+    }
+
+
+    private Cluster createCluster(CassandraKeyValueServiceConfig config, Set<InetSocketAddress> servers) {
+        long curId = cassandraId.getAndIncrement();
+        String clusterName = "cassandra" + (curId == 0 ? "" : "-" + curId);
+
+        log.info("Creating cluster {}", SafeArg.of("clusterId", clusterName));
+
+        SimpleAddressTranslator mapper = new SimpleAddressTranslator(config);
```

Comment:
Might have to switch from passing config to something else as `addressTranslator` config option uses only hostname and not hostname and port.

Reply:
Currently works without any address translation.